core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -43,7 +54,132 @@ impl RequestorNegotiationEngine {
+        let _ = self
+            .db
+            .as_dao::<EventsDao>()
+            .remove_requestor_events(subscription_id)
+            .await
+            .map_err(|e| {
+                log::warn!(
+                    "Failed to remove events related to subscription [{}].",
+                    subscription_id
+                )
+            });
+        // TODO: We could remove all resources related to Proposals.
         Ok(())
     }
+
+    pub async fn query_events(
+        &self,
+        subscription_id: &SubscriptionId,
+        timeout: f32,
+        max_events: Option<i32>,
+    ) -> Result<Vec<RequestorEvent>, QueryEventsError> {
+        let mut timeout = Duration::from_millis((1000.0 * timeout) as u64);
+        let stop_time = Instant::now() + timeout;
+        let max_events = max_events.unwrap_or(i32::max_value());
+
+        if max_events < 0 {
+            Err(QueryEventsError::InvalidMaxEvents(max_events))?
+        } else if max_events == 0 {
+            return Ok(vec![]);
+        }
+
+        loop {
+            let events = get_events_from_db(&self.db, subscription_id, max_events).await?;
+            if events.len() > 0 {
+                return Ok(events);
+            }
+
+            // Solves panic 'supplied instant is later than self'.
+            if stop_time < Instant::now() {
+                return Ok(vec![]);
+            }
+            timeout = stop_time - Instant::now();
+
+            if let Err(error) = self
+                .notifier
+                .wait_for_event_with_timeout(subscription_id, timeout)
+                .await
+            {
+                return match error {
+                    NotifierError::Timeout(_) => Ok(vec![]),
+                    NotifierError::ChannelClosed(_) => {
+                        Err(QueryEventsError::InternalError(format!("{}", error)))
+                    }
+                    NotifierError::Unsubscribed(id) => Err(QueryEventsError::Unsubscribed(id)),
+                };
+            }
+            // Ok result means, that event with required subscription id was added.
+            // We can go to next loop to get this event from db. But still we aren't sure
+            // that list won't be empty, because other query_events calls can wait for the same event.
+        }
+    }
+}
+
+async fn get_events_from_db(
+    db: &DbExecutor,
+    subscription_id: &SubscriptionId,
+    max_events: i32,
+) -> Result<Vec<RequestorEvent>, QueryEventsError> {
+    let events = db
+        .as_dao::<EventsDao>()
+        .take_requestor_events(subscription_id, max_events)
+        .await?;
+
+    // Map model events to client RequestorEvent.
+    let results = futures::stream::iter(events)
+        .then(|event| event.into_client_requestor_event(&db))
+        .collect::<Vec<Result<RequestorEvent, EventError>>>()
+        .await;
+
+    // Filter errors. Can we do something better with errors, than logging them?
```

Comment:
I'd leave logging for MVP.

btw. can you join those two iterations? It will probably shorter and much efficient bc of no need to collect in between

Reply:
I could if you propose something. Note where is the problem. First iteration makes async call. Second doesn't.
First uses futures iterator, second uses normal iterator on collection.